cdap-app-fabric/src/main/java/co/cask/cdap/runtime/monitor/RuntimeMonitor.java

```diff
@@ -0,0 +1,162 @@
+                                                      "/runtime/monitor/topics").toURL(),
+                                                 clientConfig.getAccessToken(),
+                                                 HttpURLConnection.HTTP_NOT_FOUND);
+      if (response.getResponseCode() == HttpURLConnection.HTTP_NOT_FOUND) {
+        throw new NotFoundException(response.getResponseBodyAsString());
+      }
+
+      String topicsToMonitor = response.getResponseBodyAsString(Charset.forName("UTF-8"));
+      this.topicsToMonitor = GSON.fromJson(topicsToMonitor, MAP_STRING_STRING_TYPE);
+    }
+
+    // TODO initialize from offset table for a given runtimeId
+    for (String topic : topicsToMonitor.values()) {
+      topicToRequest.put(topic, new MonitorConsumeRequest(topic, null, limit));
+    }
+  }
+
+  @Override
+  protected void run() throws Exception {
+    runThread = Thread.currentThread();
+
+    try {
+      while (!stopped) {
+        HttpResponse response = restClient.execute(HttpMethod.GET, clientConfig.getConnectionConfig().resolveURI
+                                                     (clientConfig.getApiVersion() + "/runtime/metadata").toURL(),
+                                                   GSON.toJson(topicToRequest), ImmutableMap.<String, String>of(),
+                                                   clientConfig.getAccessToken(), HttpURLConnection.HTTP_NOT_FOUND);
+
+        if (response.getResponseCode() == HttpURLConnection.HTTP_NOT_FOUND) {
+          throw new NotFoundException(response.getResponseBodyAsString());
+        }
+
+        Map<String, List<MonitorMessage>> monitorResponses =
+          GSON.fromJson(response.getResponseBodyAsString(Charset.forName("UTF-8")), MAP_STRING_MESSAGE_TYPE);
+
+        processResponse(monitorResponses);
+        Thread.sleep(pollTimeMs);
+      }
+    } catch (InterruptedException e) {
+      // Interruption means stopping the service.
+    } catch (Exception e) {
+      LOG.error("Error while fetching monitoring data from runtime {} ", runtimeId, e);
+    }
+  }
+
+  private void processResponse(Map<String, List<MonitorMessage>> monitorResponses) throws Exception {
+    String topicToMonitor = topicsToMonitor.get(Constants.AppFabric.PROGRAM_STATUS_EVENT_TOPIC);
+    publish(topicToMonitor, Constants.AppFabric.PROGRAM_STATUS_EVENT_TOPIC, monitorResponses.get(topicToMonitor));
+
+    // TODO publish metadata, lineage, workflow token
+  }
+
+  private void publish(String topic, String config, List<MonitorMessage> messages) throws Exception {
+    for (MonitorMessage message : messages) {
+      // TODO publish messages transactionally
+      MonitorConsumeRequest existing = topicToRequest.get(topic);
+      messagePublisher.publish(NamespaceId.SYSTEM.getNamespace(), cConf.get(config), message.getMessage());
+      // TODO update offset table
+      topicToRequest.put(topic, new MonitorConsumeRequest(existing.getTopic(), message.getMessageId() + 1, limit));
```

Comment:
the `+1` won't work. Message id are not number. You should structure it as part of the REST API that whether it includes the messageId or not (refer to TMS REST api, you'll see the `inclusive` parameter).